src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -36,105 +37,144 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
+                {
+                    return await HandleTokenRefreshErrorAsync(e, cachedAccessTokenItem).ConfigureAwait(false);
+                }
             }
             else
             {
-                cacheInfoTelemetry = CacheRefreshReason.ForceRefreshOrClaims;
-                logger.Info("Skipped looking for an Access Token in the cache because ForceRefresh or Claims were set. ");
+                //  No AT in the cache 
+                if (AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo != CacheRefreshReason.Expired)
+                {
+                    cacheInfoTelemetry = CacheRefreshReason.NoCachedAccessToken;
+                }
+                else
+                {
+                    cacheInfoTelemetry = CacheRefreshReason.Expired;
+                }
+
+                authResult = await GetAccessTokenAsync(false, cancellationToken, logger).ConfigureAwait(false);
+            }
+
+            AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo = cacheInfoTelemetry;
+
+            return authResult;
+        }
+
+        private async Task<AuthenticationResult> GetAccessTokenAsync(
+            bool proactivelyRefresh, 
+            CancellationToken cancellationToken, 
+            ILoggerAdapter logger)
+        {
+            await ResolveAuthorityAsync().ConfigureAwait(false);
+            MsalTokenResponse msalTokenResponse;
+            AuthenticationResult authResult;
+            MsalAccessTokenCacheItem cachedAccessTokenItem = null;
+
+            //calls to AAD
+            if (ServiceBundle.Config.AppTokenProvider == null)
+            {
+                msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+                return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
             }
 
-            if (AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo == CacheRefreshReason.NotApplicable)
+            // Bypass cache and send request to token endpoint, when 
+            // 1. Force refresh is requested, or
+            // 2. Claims are passed, or 
+            // 3. If the AT needs to be refreshed pro-actively 
+            if (_clientParameters.ForceRefresh ||
+                proactivelyRefresh || 
+                !string.IsNullOrEmpty(AuthenticationRequestParameters.Claims))
             {
-                AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo = cacheInfoTelemetry;
+                authResult = await SendTokenRequestToProviderAsync(logger, cancellationToken).ConfigureAwait(false);
+                return authResult;
             }
 
-            // No AT in the cache or AT needs to be refreshed
+            //allow only one call to the provider 
+            logger.Verbose(() => "[GetAccessTokenAsync] Entering token acquire for client credential request semaphore.");
```

Comment:
nit: I would replace `[GetAccessTokenAsync]` log prefix with `[ClientCredentialRequest]` because it's a bit more specific when looking through logs.